plugins/linguist-backend/src/processor/LinguistTagsProcessor.ts

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity, stringifyEntityRef } from '@backstage/catalog-model';
+import {
+  CatalogProcessor,
+  CatalogProcessorCache,
+} from '@backstage/plugin-catalog-node';
+import { DiscoveryService } from '@backstage/backend-plugin-api';
+import { Languages } from '@backstage/plugin-linguist-common';
+import fetch from 'node-fetch';
+import { Logger } from 'winston';
+import { HumanDuration, durationToMilliseconds } from '@backstage/types';
+
+/**
+ * A function which takes a language string returned by linguist and
+ * cleans it up prior to exporting to entity tags. If the language is emitted as
+ * the "undefined" type, then it will not be added as a tag
+ * @public
+ */
+export type TagTransformer = (tag: string) => string | undefined;
+/**
+ * A function which given an entity, determines if it should be processed for linguist tags.
+ * @public
+ */
+export type ShouldProcessEntity = (entity: Entity) => boolean;
+
+interface CachedData {
+  [key: string]: number | string[];
+  languages: string[];
+  cachedTime: number;
+}
+
+/**
+ * The constructor arguments for building the  LinguistTagsProcessor
+ * @public
+ */
+export interface LinguistTagsProcessorArgs {
+  logger: Logger;
+  discovery: DiscoveryService;
+  /**
+   * Optional transformer function which can convert a language tag prior to being
+   * attached to an entity. If the tag transformer emits an "undefined" type, then it
+   * will not be added as a tag to the entity.
+   */
+  tagTransformer?: TagTransformer;
+  /**
+   * A function which determines which entities should be processed by the LinguistTagProcessor.
+   *
+   * The default is to process all entities of kind=Component
+   */
+  shouldProcessEntity?: ShouldProcessEntity;
```

Comment:
I chose to not inline them because they are used in several places throughout the file, so a type alias made sense so that the signature can be updated in just one place if needed. Is there a particular reason you prefer in-lining?

Reply:
I thought about the API surface of the plugin / the API report. If they are "public", it is harder to modify or remove them later without a _potential_ breaking change.

Reply:
Hmm, wouldn't changing the in-line signature or fieldName of the option be just as breaking as changing anything on a type alias? Not sure I follow what the difference is. I guess the only difference is that you may change the type name and not the signature?